### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by swyxio** [breadthe/sd-buddy#6] [Ideas]
heya, so i know this project isn't open contributions but i do find the need for a GUI and i like the way you do stuff so am just penning my developing thoughts as i start to explore SD and think about what i want in a GUI

everything here is subject to the recognition that you can/should ignore any of it and i wont take any offense

1. [ ] **more cli options**. i would like a CFG parameter setting (which I think means the `--scale` param?). this one is easy, you'll probably do it eventually.  being able to tweak `--n_samples` would also be nice. and seed (max [4294967295](https://www.reddit.com/r/StableDiffusion/comments/wu1gg6/randomizing_seeds_when_running_locally/))
2. [ ] **prevent footguns**
	- [escape double quotes](https://stackoverflow.com/questions/7760545/escape-double-quotes-in-parameter#:~:text=Escape%20every%20double%20quote%20%22%20with,them%20with%20a%20caret%2C%20too.)
	- token length checker - max limit is [75](https://www.reddit.com/r/dalle2/comments/vxny8f/question_how_many_words_can_you_input_to_dalle_to/)-[77](https://www.reddit.com/r/StableDiffusion/comments/wl4cn3/the_maximum_usable_length_of_a_stable_diffusion/) tokens - prevent expensive errors
	- [errorboundary](https://twitter.com/sveltesociety/status/1326220577523314690?lang=en)? i wrote a bug and the ui went unresponsive until i opened up devtools.
3. [x] track "time elapsed" so we can see how far along we are in the generation...
	- alongside of maybe an "estimated time" thing with heuristics based on past runs and/or "complexity cost" based on the params
4. [x] seeing the results of the image in the app itself. not suuper sure how exactly to do it but its probably not hard?
5. [ ] (hardest) being able to choose the image and "refine" it - this to me was the killer UX of midjourney - this probably means storing the metadata of each image alongside the image, including a fixed seed. lets you explore a bunch of stuff in a grid and then refine it

something like that? thoughts? obv if we're aligned then i'd be happy to contribute most of these since i want them

**Comment by breadthe** [comment]
Hey Shawn, I'm somewhat familiar with your work so I really appreciate you wanting to help, and I know I can trust you with it.

For starters, I moved this to a discussion.

All these suggestions are very sensible. Some of them are on my own roadmap, others are common-sense.

1. already on the [wishlist](https://github.com/breadthe/sd-buddy#wishlist), last item. I can do this myself, unless you're quicker.
2. this one I'll leave to you... I'm not doing a lot of error handling since it's not a mission critical app but I agree these things should he handled better.
3. I was mulling the idea of showing a real-time counter while it's generating. I'll take a shot at it myself after I finish what I'm currently working on.
4. also on the [wishlist](https://github.com/breadthe/sd-buddy#wishlist) and the thing I'm currently hacking on. It's quite tricky and my solution is a bit wild, but I'm hoping it'll work. I had a working prototype last night but I found an issue with it so I'm improving it today.
5. this one is a big MAYBE, but if you can pull it off, feel free.

Now, be aware that I'm still refactoring things (extracting components, moving logic) as I see fit. At some point I want to add more structure to the app (multiple pages/sections, etc) so expect the layout and components to change.

My only request is that, if you make a PR and I've changed things in the meantime enough that there are serious merge conflicts, please pull my latest code and re-integrate your changes.

Thanks for helping!

**Reply by swyxio** [reply]
nice! ya i mean i am not sure how intensely ill be using this myself because my own local copy of stablediffusion started crashing repeatedly (i think it is out of memory but i keep restarting and it doesnt work).

**Reply by breadthe** [reply]
Are you using `--n_samples` > 1 by any chance? My SD python command crashes when I try anything > 1.

**Reply by swyxio** [reply]
yes! that was it! weird, because i initially tried n_samples up to 4 and it was fine.
